File: 백준/Silver/9012. 괄호/괄호.py
# function
def f(s):
    if len(s) == 0:
        return True
     
    fl = len(s) # 변경 전 s의 길이
    for i in range(len(s)-1):
        if s[i] + s[i+1] == '()':
            s.pop(i); s.pop(i)
            break
    if fl == len(s): # 변경 전 s와 변경 후 s의 길이 비교
        return False

    return f(s)

# main
T = int(input())

for i in range(T):
    s = list(input())
    if f(s):
        print('YES')
    else:
        print('NO')

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main() {
    int T;
    cin >> T;
    for (int i=0; i<T; i++) {
        stack<char> s;

        string str;
        cin >> str;

        bool check = true;
        for (int j=0; j<(int)str.size(); j++) {
            if (str[j] == '(') {
                s.push('(');
            } else {
                if (s.empty()) {
                    check = false;
                    break;
                } 
                s.pop();
            }
        }

        if (s.empty() && check) {
            cout << "YES" << '\n';
        } else {
            cout << "NO" << '\n';
        }
    }

    return 0;
}